### ISSUE TITLE
[Bug]: mo report  transaction error: new txn failed when running oltp delete test in push mode and deny new connection

### ISSUE THREAD

**Issue by aressu1985** [matrixorigin/matrixone#8391]
### Is there an existing issue for the same bug?

- [X] I have checked the existing issues.

### Environment

```Markdown
- Version or commit-id (e.g. v0.1.0 or 8b23a93):0129d792a1253bdf8c52afbd9ec8e239e77937d5
- Hardware parameters:
- OS type:
- Others:
```

### Actual Behavior

2023/03/13 07:50:15.366143 +0800 INFO logtail/handle.go:535 [Logtail] send block meta for "sbtest4"
2023/03/13 07:50:15.366186 +0800 INFO logtail/handle.go:538 [logtail] table meta [Insert] 272506-sbtest4: [BatchName=meta]
(attr=__mo_rowid)[26]: 12359,12360,12364
(attr=commit_time)[26]: 1678665015355527482-0,1678665015355527482-0,1678665015355527482-0
(attr=block_id)[26]: 12359,12360,12364
(attr=entry_state)[26]: false,false,false
(attr=sorted)[26]: true,true,true
(attr=%!%mo__meta_loc)[26]: 8454216b-23af-4c49-8088-3bc57082ebed:16366047_7488_7488_4:8192:16235869,8454216b-23af-4c49-8088-3bc57082ebed:16366047_7488_7488_5:8192:16235869,8454216b-23af-4c49-8088-3bc57082ebed:16366047_7488_7488_9:8192:16235869
(attr=delta_loc)[26]: ,,
(attr=committs)[26]: 1678665015355527482-0,1678665015355527482-0,1678665015355527482-0
(attr=segment_id)[26]: 1086,1086,1086

2023/03/13 07:50:15.366198 +0800 INFO logtail/handle.go:535 [Logtail] send block meta for "sbtest4"
2023/03/13 07:50:15.366210 +0800 INFO logtail/handle.go:538 [logtail] table meta [Delete] 272506-sbtest4: [BatchName=meta]
(attr=__mo_rowid)[13]: 12306,12156,12146
(attr=commit_time)[13]: 1678665015355527482-0,1678665015355527482-0,1678665015355527482-0

2023/03/13 07:50:16.559713 +0800 ERROR disttae/logtailPushModel.go:164 error: transaction error: new txn failed. please retry. {"span": {"trace_id": "77c0ff0e-c130-11ed-9008-b07b25f8b524", "kind": "statement"}}
github.com/matrixorigin/matrixone/pkg/vm/engine/disttae.(*syncLogTailTimestamp).blockUntilTxnTimeIsLegal
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/vm/engine/disttae/logtailPushModel.go:164
github.com/matrixorigin/matrixone/pkg/vm/engine/disttae.(*Engine).New
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/vm/engine/disttae/engine.go:377
github.com/matrixorigin/matrixone/pkg/vm/engine.(*EntireEngine).New
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/vm/engine/entire_engine.go:33
github.com/matrixorigin/matrixone/pkg/frontend.(*TxnHandler).NewTxn
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/session.go:1562
github.com/matrixorigin/matrixone/pkg/frontend.(*Session).TxnStart
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/session.go:1110
github.com/matrixorigin/matrixone/pkg/frontend.(*TxnHandler).GetTxn
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/session.go:1723

github.com/matrixorigin/matrixone/pkg/frontend.(*MysqlCmdExecutor).doComQuery
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/mysql_cmd_executor.go:3107
github.com/matrixorigin/matrixone/pkg/frontend.(*MysqlCmdExecutor).ExecRequest
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/mysql_cmd_executor.go:3889
github.com/matrixorigin/matrixone/pkg/frontend.(*Routine).handleRequest
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/routine.go:175
github.com/matrixorigin/matrixone/pkg/frontend.(*RoutineManager).Handler
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/routine_manager.go:320
github.com/fagongzi/goetty/v2.(*server).doConnection
        /home/github/go/pkg/mod/github.com/fagongzi/goetty/v2@v2.0.3-0.20221212132037-abf2d4c05484/application.go:381
github.com/fagongzi/goetty/v2.(*server).doStart.func2.1
        /home/github/go/pkg/mod/github.com/fagongzi/goetty/v2@v2.0.3-0.20221212132037-abf2d4c05484/application.go:344
2023/03/13 07:50:16.559800 +0800 ERROR frontend/util.go:424 GetTxn. error:transaction error: new txn failed. please retry. connectionId 12644
github.com/matrixorigin/matrixone/pkg/frontend.logErrorf
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/util.go:424
github.com/matrixorigin/matrixone/pkg/frontend.(*TxnHandler).GetTxn
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/session.go:1725
github.com/matrixorigin/matrixone/pkg/frontend.doUse
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/mysql_cmd_executor.go:664
github.com/matrixorigin/matrixone/pkg/frontend.(*MysqlCmdExecutor).handleChangeDB
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/mysql_cmd_executor.go:682
github.com/matrixorigin/matrixone/pkg/frontend.(*MysqlCmdExecutor).doComQuery
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/mysql_cmd_executor.go:3107
github.com/matrixorigin/matrixone/pkg/frontend.(*MysqlCmdExecutor).ExecRequest
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/mysql_cmd_executor.go:3889
github.com/matrixorigin/matrixone/pkg/frontend.(*Routine).handleRequest
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/routine.go:175
github.com/matrixorigin/matrixone/pkg/frontend.(*RoutineManager).Handler
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/routine_manager.go:320
github.com/fagongzi/goetty/v2.(*server).doConnection
        /home/github/go/pkg/mod/github.com/fagongzi/goetty/v2@v2.0.3-0.20221212132037-abf2d4c05484/application.go:381

and then , the new connection was denied.

<img width="893" alt="image" src="https://user-images.githubusercontent.com/47846308/224604397-f25c92df-9beb-4035-892e-cde2b829c9cb.png">

**Comment by m-schen** [comment]
the error was return when we try to open a new txn, we assume that txn time is T and CN will block it after we have received log tail (t >= T).
If the block time was more than 1 minute ( I write it at code `maxTimeToNewTransaction`, and I'm not sure if it was a good time), will return the error 'new txn failed.'

**Comment by m-schen** [comment]
@cnutshell can help find out the specific reasons why CN hasn't receive the log tail (which t >= T).

**Comment by cnutshell** [comment]
It seems that the mo-service wasn't fully startup.
I couldn't reproduce it in my local environment. 
Can you @aressu1985 please describe the detailed context information?
Or how can I find the whole running log?

**Comment by cnutshell** [comment]
<img width="1361" alt="image" src="https://user-images.githubusercontent.com/20291742/224651712-3bd1bce9-48f1-4716-9469-19b0216d27e6.png">
The first panic from logtail client side, as a result of nil `LogtailClient.limiter`. Its time is `07:45:40.380022`.
Hard to reason about it, if `LogtailClient` was initialized successfully, `LogtailClient.limiter` would be nil.
Maybe some concurrent bug with the initialization of `LogtailClient`.

**Comment by cnutshell** [comment]
```bash
2023/03/13 07:47:06.406410 +0800 ERROR disttae/logtailPushModel.go:164 error: transaction error: new txn failed. please retry. {"span": {"trace_id": "fefd1d92-c12f-11ed-9008-b07b25f8b524", "kind": "statement"}}
2023/03/13 07:47:06.406510 +0800 ERROR frontend/util.go:424 GetTxn. error:transaction error: new txn failed. please retry. connectionId 6359
```

The first `new txn failed` took place at time `07:47:06.406410`, I've checked the logic for `types.Engine.subscriber`, maybe a concurrency bug for its using and reconstruction. 

I'll transfer this issue to @m-schen.

**Comment by m-schen** [comment]
I will do a check for reconnect process.

**Comment by m-schen** [comment]
先解决手头上的s-1后做这个，目前的问题是re-connect时与其他并发的查询会race导致了panic.

**Comment by m-schen** [comment]
reconnect race. 后面会解决。

**Comment by m-schen** [comment]
I have fix the reconnect race at #8725

**Comment by m-schen** [comment]
fix a reconnect hung today.  just look today's daily run perform.

**Comment by aressu1985** [comment]
This is not fixed for commit: f5e458d596b37ce62587bb876401f008010b1adb

2023/04/04 10:13:51.653064 +0800 INFO db/scannerop.go:370 Mergeblocks stats: map[272467:row91925[2023-04-04 10:16:03.763482949 +0800 CST m=+391.691757840] 272468:row91871[2023-04-04 10:16:03.768282619 +0800 CST m=+391.696557506] 272469:row91879[2023-04-04 10:16:03.767680783 +0800 CST m=+391.695955670] 272470:row91899[2023-04-04 10:16:03.766056811 +0800 CST m=+391.694331698] 272471:row91919[2023-04-04 10:16:03.764415263 +0800 CST m=+391.692690148] 272472:row91953[2023-04-04 10:16:03.761578342 +0800 CST m=+391.689853229] 272473:row91916[2023-04-04 10:16:03.7648327 +0800 CST m=+391.693107586] 272474:row91753[2023-04-04 10:16:03.77891651 +0800 CST m=+391.707191397] 272475:row91935[2023-04-04 10:16:03.763362561 +0800 CST m=+391.691637447] 272476:row91861[2023-04-04 10:16:03.769799655 +0800 CST m=+391.698074542]]
2023/04/04 10:13:51.653152 +0800 INFO db/scannerop.go:376 Mergeblocks available mem: 223g
2023/04/04 10:13:52.008385 +0800 ERROR disttae/logtailPushModel.go:126 new txn failed because lack of enough log tail. txn time is [PhysicalTime:1680574370070651231 NodeID:33571 ]
github.com/matrixorigin/matrixone/pkg/vm/engine/disttae.(*pushClient).checkTxnTimeIsLegal
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/vm/engine/disttae/logtailPushModel.go:126
github.com/matrixorigin/matrixone/pkg/vm/engine/disttae.(*Engine).New
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/vm/engine/disttae/engine.go:350
github.com/matrixorigin/matrixone/pkg/vm/engine.(*EntireEngine).New
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/vm/engine/entire_engine.go:33
github.com/matrixorigin/matrixone/pkg/frontend.(*TxnHandler).NewTxn
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/txn.go:178
github.com/matrixorigin/matrixone/pkg/frontend.(*Session).TxnCreate
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/txn.go:476
github.com/matrixorigin/matrixone/pkg/frontend.(*TxnHandler).GetTxn
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/txn.go:362
github.com/matrixorigin/matrixone/pkg/frontend.(*TxnCompilerContext).GetSubscriptionMeta
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/compiler_context.go:781
github.com/matrixorigin/matrixone/pkg/frontend.(*TxnCompilerContext).ensureDatabaseIsNotEmpty
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/compiler_context.go:254
github.com/matrixorigin/matrixone/pkg/frontend.(*TxnCompilerContext).Resolve
        /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/frontend/compiler_context.go:276

<img width="1123" alt="image" src="https://user-images.githubusercontent.com/47846308/229678457-1c402f10-f7a3-472a-9138-e4ae5265c981.png">

**Comment by zhangxu19830126** [comment]
I started working on this issue

**Comment by zhangxu19830126** [comment]
Opened profile to run sysbench locally for 2 hours and did not reproduce the issue. Keep running tomorrow, once it appears, we can use profile to confirm if there is a deadlock and what is blocking

**Comment by zhangxu19830126** [comment]
The test ran continuously for several days without recurrence. First mark it as holding and look at the message analysis when it appears.

**Comment by zhangxu19830126** [comment]
[hung.log](https://github.com/matrixorigin/matrixone/files/11241221/hung.log)

**Comment by zhangxu19830126** [comment]
The execution is blocked in the batch_processor.go @xzxiong

**Comment by xzxiong** [comment]
> [hung.log](https://github.com/matrixorigin/matrixone/files/11241221/hung.log)

the following goroutine is directly reason for the hang:
- goroutine 150, want to allocBuffer
     - wait goroutine 162 to  release its buffer
- goroutine 162, want to Send info to the MOCollector 
    - wait all goroutines, like 150, to accept.

```
goroutine 150 [sync.Cond.Wait, 250 minutes]:
sync.runtime_notifyListWait(0xc000800010, 0x106)
    /usr/local/go/src/runtime/sema.go:517 +0x14c
sync.(*Cond).Wait(0xc000517e28?)
    /usr/local/go/src/sync/cond.go:70 +0x8c
github.com/matrixorigin/matrixone/pkg/util/export.(*MOCollector).allocBuffer(0xc000830000)
    /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/util/export/batch_processor.go:354 +0x3e
github.com/matrixorigin/matrixone/pkg/util/export.(*bufferHolder).getBuffer(0xc002498090)
    /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/util/export/batch_processor.go:110 +0x2b
github.com/matrixorigin/matrixone/pkg/util/export.(*bufferHolder).Add(0xc002498090, {0x2cbd6c0, 0xcdd2e12360})
    /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/util/export/batch_processor.go:133 +0x9e
github.com/matrixorigin/matrixone/pkg/util/export.(*MOCollector).doCollect(0xc000830000, 0x0?)
    /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/util/export/batch_processor.go:396 +0x265
created by github.com/matrixorigin/matrixone/pkg/util/export.(*MOCollector).Start
    /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/util/export/batch_processor.go:336 +0x13e
```

```
goroutine 162 [select, 250 minutes]:
github.com/matrixorigin/matrixone/pkg/util/export.(*MOCollector).Collect(0x0?, {0x2cf4fb8, 0xc000475230}, {0x2cbd6c0?, 0xc3993faea0?})
    /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/util/export/batch_processor.go:311 +0x8c
github.com/matrixorigin/matrixone/pkg/util/trace/impl/motrace.(*batchSpanProcessor).OnEnd(0xc000800200, {0x2ceb0a0?, 0xc3993faea0?})
    /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/util/trace/impl/motrace/batch_span_processor.go:65 +0xbc
github.com/matrixorigin/matrixone/pkg/util/trace/impl/motrace.(*MOSpan).End(0xc3993faea0, {0x0?, 0x1a923e6?, 0x0?})
    /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/util/trace/impl/motrace/mo_trace.go:152 +0x109
github.com/matrixorigin/matrixone/pkg/util/trace/impl/motrace.(*itemBuffer).GetBatch(0xc0b7698180, {0x2cf4fb8?, 0xc000e46090?}, 0xc6559f0f0c?)
    /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/util/trace/impl/motrace/buffer_pipe.go:252 +0x1d8
github.com/matrixorigin/matrixone/pkg/util/export.(*bufferGenerateReq).handle(0xc735836ff0, 0xc750775f10?)
    /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/util/export/batch_processor.go:152 +0x3f
github.com/matrixorigin/matrixone/pkg/util/export.(*MOCollector).doGenerate(0xc000830000, 0x2e30303433203a09?)
    /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/util/export/batch_processor.go:436 +0x176
created by github.com/matrixorigin/matrixone/pkg/util/export.(*MOCollector).Start
    /data1/action-runner/_work/mo-nightly-regression/mo-nightly-regression/head/pkg/util/export/batch_processor.go:340 +0x1bc
```

**Comment by xzxiong** [comment]
@aressu1985 ptal, i have merged the bugfix yesterday.

**Comment by aressu1985** [comment]
seed fixed commit: a181a764b
